### PR TITLE
Q-Criterion Proxy Feature: vortex zone indicator for p_tan/p_oodc

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**Physics-motivated input feature targeting p_tan and p_oodc.** The Q-criterion in fluid mechanics (Q = 0.5*(||Ω||² - ||S||²)) distinguishes vortex-dominated regions (Q > 0: wake vortices, tandem slot, separation) from strain-dominated regions (Q < 0: freestream, attached BL). The model currently has no way to identify which nodes are in vortex-dominated wake zones — it must infer this from DSDF distance alone.

We compute a **Q-criterion proxy** from existing input features at inference time (no CFD data required):
- The cross-product of the DSDF gradient direction with the freestream velocity direction approximates the local signed vorticity near the airfoil surface
- Weighted by `exp(-dsdf * 3.0)` to decay away from the surface (where boundary-layer vorticity is physically dominant)
- This gives a per-node scalar that is positive in upper-surface wake zones, negative below the foil, and near-zero in the freestream

For tandem configs: the slot between fore and aft foils is a strong vortex zone (fore-TE trailing vortex impinging on aft-foil suction surface). This is exactly where p_tan errors are concentrated. An explicit vortex-zone indicator should help the model partition its capacity between attached-flow and wake-interaction regimes.

**Why it follows the winning pattern:** Every durable gain has come from physics-motivated input features. This is purely feature engineering (no architecture change, no loss change). Never tried in any of 1903 experiments.

**Target metrics:** p_tan (biggest gap: 27.87 vs ensemble 29.1), p_oodc (7.64 vs ensemble 6.6)

**Literature support:** RITA paper (arXiv:2504.06758) shows local flow invariants (Q, R, velocity gradient tensor) classify flow regime from geometry features alone with high accuracy.

### Step 1: Understand the input feature structure

Read `prepare_multi.py` and `train.py` to identify:
- The channel index of DSDF for foil 1 (scalar distance, channel ~0 in DSDF group)
- The channel indices of DSDF gradient x,y for foil 1 (channels ~1,2 in DSDF group)
- The channel indices of DSDF scalar and gradient for foil 2
- The AoA channel index

### Step 2: Implement the Q-criterion proxy feature

Add `--q_criterion_feature` flag. When enabled, compute the proxy from existing input channels and append to `x`:

```python
def compute_q_criterion_proxy(x, pos, args):
    """Compute vorticity proxy from DSDF gradient x freestream direction.
    
    Returns: [B, N, 2] — q_proxy for foil1 and foil2 (tandem) or [foil1, 0] for single.
    """
    # VERIFY THESE CHANNEL INDICES by reading prepare_multi.py output feature order
    # Typical layout: dsdf1 (dist), dsdf1_grad_x, dsdf1_grad_y, dsdf2 (dist), dsdf2_grad_x, dsdf2_grad_y
    dsdf1_dist    = x[:, :, DSDF1_DIST_CH]      # [B, N] distance to foil 1
    dsdf1_grad_x  = x[:, :, DSDF1_GRADX_CH]     # [B, N] DSDF grad x, foil 1
    dsdf1_grad_y  = x[:, :, DSDF1_GRADY_CH]     # [B, N] DSDF grad y, foil 1
    dsdf2_dist    = x[:, :, DSDF2_DIST_CH]      # [B, N] distance to foil 2
    dsdf2_grad_x  = x[:, :, DSDF2_GRADX_CH]     # [B, N] DSDF grad x, foil 2
    dsdf2_grad_y  = x[:, :, DSDF2_GRADY_CH]     # [B, N] DSDF grad y, foil 2
    aoa_val       = x[:, 0, AOA_CH]             # [B] AoA in whatever units stored
    
    # Convert AoA to radians if stored in degrees (check units in prepare_multi.py)
    aoa_rad = aoa_val * (np.pi / 180.0) if aoa_in_degrees else aoa_val
    
    # Freestream direction vector [cos(AoA), sin(AoA)]
    u_inf_x = torch.cos(aoa_rad).unsqueeze(1)   # [B, 1]
    u_inf_y = torch.sin(aoa_rad).unsqueeze(1)   # [B, 1]
    
    # Cross product: dsdf_grad × u_inf = dsdf_grad_x * u_inf_y - dsdf_grad_y * u_inf_x
    # Positive = upper-surface/wake zone, Negative = lower-surface zone
    vort_proxy_foil1 = dsdf1_grad_x * u_inf_y - dsdf1_grad_y * u_inf_x  # [B, N]
    vort_proxy_foil2 = dsdf2_grad_x * u_inf_y - dsdf2_grad_y * u_inf_x  # [B, N]
    
    # Weight by surface proximity: decays exponentially away from surface
    # Use dsdf distance (normalized) — smaller dsdf = closer to surface = stronger BL vorticity
    q_proxy_foil1 = vort_proxy_foil1 * torch.exp(-dsdf1_dist.abs() * 3.0)  # [B, N]
    q_proxy_foil2 = vort_proxy_foil2 * torch.exp(-dsdf2_dist.abs() * 3.0)  # [B, N]
    
    # For single-foil samples, foil2 DSDF is typically zero — q_proxy_foil2 ≈ 0 naturally
    # Stack as 2 new channels
    q_features = torch.stack([q_proxy_foil1, q_proxy_foil2], dim=-1)  # [B, N, 2]
    
    return q_features
```

Then append to `x` before the forward pass:
```python
if args.q_criterion_feature:
    q_feat = compute_q_criterion_proxy(x, pos, args)
    x = torch.cat([x, q_feat], dim=-1)  # [B, N, C+2]
```

**CRITICAL CHECK before training:** On a single batch, verify:
1. Print q_proxy values for surface nodes — should show positive values on upper surface, negative on lower surface for a positive AoA sample
2. Print q_proxy for wake nodes (far downstream) — should decay toward zero
3. For a tandem sample: check that foil2 q_proxy also has non-zero values in the slot region

If the signs look wrong (all positive or all negative everywhere), check the AoA sign convention in your data.

### Step 3: Model input dimension

Since we add 2 new input channels, the model's input dimension changes. Ensure the embedding layer handles the larger input:
- If using `--n_layers 3 --slice_num 96` with default input dim, the model auto-detects input size from data
- Verify the model initializes correctly with the new feature by checking the first forward pass doesn't error

### Step 4: Training commands

```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent tanjiro --seed 42 \
  --wandb_name "tanjiro/q-criterion-proxy-s42" \
  --wandb_group "q-criterion-proxy-feature" \
  --q_criterion_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same but CUDA_VISIBLE_DEVICES=1, --seed 73, --wandb_name "tanjiro/q-criterion-proxy-s73"
```

## Baseline

Current baseline (PR #2290, Re-Stratified Sampling, merged 2026-04-08, W&B-verified):

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in   | 11.742 | < 11.74 |
| p_oodc | 7.643  | < 7.64 |
| **p_tan** | **27.874** | < 27.87 (PRIMARY TARGET) |
| p_re   | 6.419  | < 6.42 |

W&B baseline runs: k5qwvce4 (seed 42), 7oa5xfhi (seed 73). Group: `re-stratified-sampling`.

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Q-Criterion Proxy Feature

### Setup
- **Hypothesis**: DSDF gradient × freestream direction as signed vorticity proxy, weighted by surface proximity decay. Adds 2 channels (foil-1 and foil-2 Q-proxies).
- **Flag**: `--q_criterion_feature`
- **Also includes**: `--re_stratified_sampling --re_extreme_weight 2.0` (baseline flags from PR #2290)

### Results (surface MAE pressure)

| Metric | Baseline (PR #2290) | Seed 42 | Seed 73 | Avg | Delta |
|--------|-------------------|---------|---------|-----|-------|
| p_in   | **11.742** | 12.6 ❌ | 11.5 ✅ | 12.05 | +2.6% |
| p_oodc | **7.643** | 7.7 ❌ | 7.9 ❌ | 7.8 | +2.1% |
| p_tan  | **27.874** | 27.5 ✅ | 28.5 ❌ | 28.0 | +0.5% |
| p_re   | **6.419** | 6.4 ✅ | 6.5 ❌ | 6.45 | +0.5% |

### W&B Runs
- Seed 42: `ila03wkz` — [view run](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ila03wkz)
- Seed 73: `6xij9aap` — [view run](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6xij9aap)
- Group: `q-criterion-proxy-feature`

### Analysis
- **Seed-level variance is high** — results swing between seeds (seed 42 beats p_tan at 27.5 while seed 73 beats p_in at 11.5), but on average no consistent improvement.
- **Seed 42 p_tan=27.5** is the best single-seed p_tan result we've seen, beating baseline by 1.3%. This suggests the vortex proxy may help the model partition its capacity for tandem samples. But the gain doesn't replicate across seeds.
- **Seed 73 p_in=11.5** is also strong (vs baseline 11.742), but paired with worse p_tan=28.5.
- The Q-criterion proxy feature doesn't hurt catastrophically (unlike mirror augmentation), but doesn't reliably improve either.

### Suggested Follow-ups
- The feature shows promise for p_tan specifically — the vortex zone indicator helps with tandem slot interactions in some runs. A variation could be: only apply the Q-proxy to tandem samples, or weight it by gap/stagger.
- The DSDF channel interpretation (which channels are gradients vs distances) may need further verification — if channels 4-5 aren't actually gradient (x,y) for foil-1, the cross-product is meaningless.
- Consider a learned vortex indicator (small MLP over DSDF+position features) instead of the hand-crafted cross-product formula.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2307 because ADVISOR: Closing — no metric improved on average. The high seed variance (p_tan: 27.5 vs 28.5 across seeds) indicates the Q-criterion proxy is not providing a reliable signal.

**Interesting data point:** Seed 42's p_tan=27.5 is the best single-seed tandem result we've seen (-1.3% vs baseline). But it doesn't replicate with seed 73. This suggests the vortex zone information might be marginally useful for tandem, but the hand-crafted cross-product formula (DSDF grad × freestream) isn't a robust encoding.

**Pattern update:** This is our 9th consecutive failure since PR #2290. We're in full plateau territory. The input feature space AND simple loss modifications are saturated. The next round will focus on fundamentally different approaches — training dynamics, data-centric methods, and novel architectural mechanisms.

New assignment coming shortly..
